### DISCUSSION TITLE
第5节课作业(2班)

### DISCUSSION THREAD

**Discussion by MING-ZCH** [InternLM/Tutorial#323] [作业]
# 在下方讨论区提交作业（图片/链接形式均可）~
## 基础作业：

- 使用 LMDeploy 以本地对话、网页Gradio、API服务中的一种方式部署 InternLM-Chat-7B 模型，生成 300 字的小故事（需截图）

## 进阶作业（可选做）：

- 将第四节课训练自我认知小助手模型使用 LMDeploy 量化部署到 OpenXLab 平台。
- 对internlm-chat-7b模型进行量化，并同时使用KV Cache量化，使用量化后的模型完成API服务的部署，分别对比模型量化前后和 KV Cache 量化前后的显存大小（将 bs设置为 1 和 max len 设置为512）。  
- 在自己的任务数据集上任取若干条进行Benchmark测试，测试方向包括：  
（1）TurboMind推理+Python代码集成  
（2）在（1）的基础上采用W4A16量化  
（3）在（1）的基础上开启KV Cache量化  
（4）在（2）的基础上开启KV Cache量化  
（5）使用Huggingface推理

⭐备注：**由于进阶作业较难，完成基础作业之后就可以先提交作业了，在后续的大作业项目中使用这些技术将作为重要的加分点！**

**Comment by DarthThomas** [comment]
### 基础作业
![截屏2024-01-15 22 23 37](https://github.com/InternLM/tutorial/assets/20266717/4bf5f527-d83d-4529-b422-284b24bb64ba)

**Comment by kyrie2to11** [comment]
[InterLM L-5: LMDeploy 大模型量化部署实践](https://zhuanlan.zhihu.com/p/678012554) 基础作业打卡~

**Comment by xulinui** [comment]
https://h90ag9106t.feishu.cn/docx/X8eudSxnBowylDxETKcc75JVnyb

**Comment by ZixinxinWang** [comment]
基础作业：
利用api:
<img width="590" alt="image" src="https://github.com/InternLM/tutorial/assets/152623177/9ea151e6-8a88-404a-ac3c-3e82d44efdd2">

**Comment by CoderJackZhu** [comment]
基础作业：
使用api：
![image](https://github.com/InternLM/tutorial/assets/42195616/43ca8c25-ec3e-4e42-b426-9da00aa24083)
使用gradio前端
![image](https://github.com/InternLM/tutorial/assets/42195616/cd13afb6-d3b1-472d-a4e3-7244df5acb5f)
进阶作业（可选做）：
量化后的显存：
![image](https://github.com/InternLM/tutorial/assets/42195616/c50a098c-70e8-454f-9233-b441dfcdbeec)

**Comment by pranerd** [comment]
基础作业：
使用本地对话：
![internlm_lmdeploy_1](https://github.com/InternLM/tutorial/assets/19358928/dd993453-680d-48a7-aa6a-451ebc169a83)
使用API服务：
![internlm_lmdeploy_2](https://github.com/InternLM/tutorial/assets/19358928/c8e62260-c643-47db-8a66-090fced788ef)
![internlm_lmdeploy_3](https://github.com/InternLM/tutorial/assets/19358928/64549927-ca0f-4274-9590-65ae9fc0e191)
![internlm_lmdeploy_4](https://github.com/InternLM/tutorial/assets/19358928/9f8897d5-022b-4307-9944-e84daa23b54f)
使用gradio
gradio想到相当于client，有两种处理方式
这里直接启动作为前端的 Gradio
###  Gradio+ApiServer。必须先开启 Server，此时 Gradio 为 Client
```
lmdeploy serve gradio http://0.0.0.0:23333 \
	--server_name 0.0.0.0 \
	--server_port 6006 \
	--restful_api True
```
![internlm_lmdeploy_7](https://github.com/InternLM/tutorial/assets/19358928/b596a73d-7d98-47c3-961f-255ce093a48a)
![internlm_lmdeploy_8](https://github.com/InternLM/tutorial/assets/19358928/6515a190-29d4-4e90-aa4c-b20a858a41d3)

也可以直接启动 Gradio，此时没有 API Server，TurboMind 直接与 Gradio 通信。具体命令为：
`lmdeploy serve gradio ./workspace`
![internlm_lmdeploy_5](https://github.com/InternLM/tutorial/assets/19358928/e224b1d6-15fb-4ffa-8acd-9500f2d5982a)
![internlm_lmdeploy_6](https://github.com/InternLM/tutorial/assets/19358928/0721a1f6-61cd-45fd-acf4-21c651d9e55e)

**Comment by RyanWu31** [comment]
https://github.com/RyanWu31/tutorial_notebook/blob/main/README.md

**Comment by dcuplover** [comment]
![5-1](https://github.com/InternLM/tutorial/assets/68799899/e7fd012c-3317-4a90-96d2-bfb66585146c)

**Comment by songyuc** [comment]
## 基础作业：
-   使用 LMDeploy 以本地对话、网页Gradio、API服务中的一种方式部署 InternLM-Chat-7B 模型，生成 300 字的小故事


![image](https://github.com/InternLM/tutorial/assets/27288110/7e393252-4d69-4faf-a05d-4a2b0ee041a8)

**Comment by YanxingLiu** [comment]
基础作业：
使用 LMDeploy 以本地对话、网页Gradio、API服务中的一种方式部署 InternLM-Chat-7B 模型，生成 300 字的小故事
<img width="1300" alt="image" src="https://github.com/InternLM/tutorial/assets/42299757/6d6c384b-3829-4ba5-bfca-1cf960ec09c6">

**Comment by zyy0831** [comment]
使用 LMDeploy 以API服务方式部署 InternLM-Chat-7B 模型，生成 300 字的小故事
![image](https://github.com/InternLM/tutorial/assets/43286447/0784d0ef-afe1-45ec-ad4c-3e24dbca18cd)

**Comment by Mr-Vincenyang** [comment]
基础作业：
使用 LMDeploy 以本地对话、网页Gradio、API服务中的一种方式部署 InternLM-Chat-7B 模型，生成 300 字的小故事
![image](https://github.com/InternLM/tutorial/assets/120323418/1744abab-2903-416b-b1dd-39f90712634a)

**Comment by PrinceRay7** [comment]
第五节课作业
https://github.com/PrinceRay7/InternLM_HW/blob/main/hw5/homework/homework.md

**Comment by ljqiff** [comment]
基础作业：
使用 LMDeploy 以本地对话、网页Gradio、API服务中的一种方式部署 InternLM-Chat-7B 模型，生成 300 字的小故事（需截图）

本地：
![2](https://github.com/InternLM/tutorial/assets/29425919/a853564b-b824-4510-8d57-9d6d390a6b9f)
网页Gradio：
![3](https://github.com/InternLM/tutorial/assets/29425919/8d8d314b-3ea6-41c7-9670-3ffb9740175c)
API：
![4](https://github.com/InternLM/tutorial/assets/29425919/8014ca21-d806-443d-b1ba-11ce1b04b107)

进阶作业（可选做）：
将第四节课训练自我认知小助手模型使用 LMDeploy 量化部署到 OpenXLab 平台。
部署：
MoelScope 平台。
https://www.modelscope.cn/models/ljqiff/learn4class/summary
OpenXlab
https://openxlab.org.cn/apps/detail/ljqiff/CarLLMTest
对internlm-chat-7b模型进行量化，并同时使用KV Cache量化，使用量化后的模型完成API服务的部署，分别对比模型量化前后和 KV Cache 量化前后的显存大小（将 bs设置为 1 和 max len 设置为512）。
在自己的任务数据集上任取若干条进行Benchmark测试，测试方向包括：
（1）TurboMind推理+Python代码集成
（2）在（1）的基础上采用W4A16量化

（3）在（1）的基础上开启KV Cache量化
（4）在（2）的基础上开启KV Cache量化
（5）使用Huggingface推理

**Comment by kishiokon** [comment]
![4TECRT ZJ~GP{R@L_Y U$ZD](https://github.com/InternLM/tutorial/assets/92700495/cd0ceaac-64b1-42d5-9b74-6567ba785f43)
本地对话形式生成小故事

**Comment by kaoyansoft123** [comment]
基础作业：生成300字小故事
![image](https://github.com/InternLM/tutorial/assets/152302734/28995db5-6caf-408d-9901-0e35fd8cfd5a)
1.TurboMind 推理+命令行本地对话
![image](https://github.com/InternLM/tutorial/assets/152302734/be63fa36-d971-4469-9a07-5abd69fc2e37)
2.TurboMind推理+API服务
![image](https://github.com/InternLM/tutorial/assets/152302734/070b6988-fd49-479d-802a-485d34dc6dd3)
![image](https://github.com/InternLM/tutorial/assets/152302734/82fae57d-ec2a-427c-ba76-27c4c9cf91ec)
![image](https://github.com/InternLM/tutorial/assets/152302734/9e102811-a5ba-474b-92ca-09f0cf90e427)
3.TurboMind 服务作为后端
![image](https://github.com/InternLM/tutorial/assets/152302734/187bbc21-d3db-4d4d-8f9e-26b4488eccbf)
4.TurboMind 推理作为后端
![image](https://github.com/InternLM/tutorial/assets/152302734/deb90ae9-c9d9-49af-b635-d2a1faf8068a)
量化比较：
1.W4A16 量化：
![image](https://github.com/InternLM/tutorial/assets/152302734/0b5807b6-4ada-418a-baed-257d201bb50d)
2.KV Cache 量化+W4A16 量化
![image](https://github.com/InternLM/tutorial/assets/152302734/354fe5c8-ec1c-492c-8fac-53ef1bac10a4)
